pandas/tests/series/test_analytics.py

```diff
@@ -225,101 +225,159 @@ def test_median(self):
-                    '2 min', '2 min'], dtype='timedelta64[ns]')
-        tm.assert_series_equal(s.mode(), exp)
-
-        # Test mixed dtype.
-        exp = Series(['foo'])
-        s = Series([1, 'foo', 'foo'])
-        tm.assert_series_equal(s.mode(), exp)
-
+                    '2 min', '2 min', 'nan', 'nan'],
+                   dtype='timedelta64[ns]')
+        expected2 = Series(expected2, dtype='timedelta64[ns]')
+        tm.assert_series_equal(s.mode(dropna), expected2)
+
+    @pytest.mark.parametrize('dropna, expected1, expected2, expected3', [
+        (True, Categorical([1, 2], categories=[1, 2]),
+         Categorical(['a'], categories=[1, 'a']),
+         Categorical([1, 3], categories=[1, 2, 3])),
+        (False, Categorical([np.nan], categories=[1, 2]),
+         Categorical([np.nan, 'a'], categories=[1, 'a']),
+         Categorical([np.nan, 1, 3], categories=[1, 2, 3])),
+    ])
+    def test_mode_category(self, dropna, expected1, expected2, expected3):
+        s = Series(Categorical([1, 2, np.nan, np.nan]))
+        expected1 = Series(expected1, dtype='category')
+        tm.assert_series_equal(s.mode(dropna), expected1)
+
+        s = Series(Categorical([1, 'a', 'a', np.nan, np.nan]))
+        expected2 = Series(expected2, dtype='category')
+        tm.assert_series_equal(s.mode(dropna), expected2)
+
+        s = Series(Categorical([1, 1, 2, 3, 3, np.nan, np.nan]))
+        expected3 = Series(expected3, dtype='category')
+        tm.assert_series_equal(s.mode(dropna), expected3)
+
+    @pytest.mark.parametrize('dropna, expected1, expected2', [
+        (True, [2**63], [1, 2**63]),
+        (False, [2**63], [1, 2**63])
+    ])
+    def test_mode_intoverflow(self, dropna, expected1, expected2):
         # Test for uint64 overflow.
-        exp = Series([2**63], dtype=np.uint64)
+        expected1 = Series(expected1, dtype=np.uint64)
         s = Series([1, 2**63, 2**63], dtype=np.uint64)
-        tm.assert_series_equal(s.mode(), exp)
+        tm.assert_series_equal(s.mode(dropna), expected1)
 
-        exp = Series([1, 2**63], dtype=np.uint64)
+        expected2 = Series(expected2, dtype=np.uint64)
         s = Series([1, 2**63], dtype=np.uint64)
-        tm.assert_series_equal(s.mode(), exp)
-
-        # Test category dtype.
-        c = Categorical([1, 2])
-        exp = Categorical([1, 2], categories=[1, 2])
-        exp = Series(exp, dtype='category')
-        tm.assert_series_equal(Series(c).mode(), exp)
-
-        c = Categorical([1, 'a', 'a'])
-        exp = Categorical(['a'], categories=[1, 'a'])
-        exp = Series(exp, dtype='category')
-        tm.assert_series_equal(Series(c).mode(), exp)
-
-        c = Categorical([1, 1, 2, 3, 3])
-        exp = Categorical([1, 3], categories=[1, 2, 3])
-        exp = Series(exp, dtype='category')
-        tm.assert_series_equal(Series(c).mode(), exp)
+        tm.assert_series_equal(s.mode(dropna), expected2)
+
+    @pytest.mark.skipif(not compat.PY3, reason="only PY3")
```

Comment:
It seems that in Python 3 a warning is raised if the resulting modes cannot be sorted but there is no warning in Python 2:

Python 3
```
In [2]: s = pd.Series([1, 'foo', 'foo', np.nan, np.nan])                       
In [3]: s.mode(False).sort_values().reset_index(drop=True)
Out[3]:
/home/paul/pandas-reidy-p/pandas/core/algorithms.py:840: UserWarning: Unable                                    
 to sort modes: '<' not supported between instances of 'float' and 'str'  
  warn("Unable to sort modes: {error}".format(error=e))                                                             
0    foo
1    NaN
dtype: object  
```

Python 2
```
In [2]: s = pd.Series([1, 'foo', 'foo', np.nan, np.nan])                       
In [3]: s.mode(False).sort_values().reset_index(drop=True)
Out[3]:
0    foo
1    NaN
dtype: object  
```

Relevant source code:
https://github.com/pandas-dev/pandas/blob/b3f07b2551b7eb40408ca28416ca5e21f8fb37b4/pandas/core/algorithms.py#L833-L838